### ISSUE TITLE
peak_pick corner cases

### ISSUE THREAD

**Issue by bmcfee** [bmcfee/librosa#161]
I'm writing correctness tests for the peak picker, based on the guarantees described in the docstring:

    1. `x[n] == max(x[n - pre_max:n + post_max])`
    2. `x[n] >= mean(x[n - pre_avg:n + post_avg]) + delta`
    3. `n - previous_n > wait`

The parameters sections describes the valid inputs as follows:

    x         : np.ndarray [shape=(n,)]
    pre_max   : int >= 0 [scalar]
    post_max  : int >= 0 [scalar]
    pre_avg   : int >= 0 [scalar]
    post_avg  : int >= 0 [scalar]
    delta     : float >= 0 [scalar]
    wait      : int >= 0 [scalar]

From this description, `pre_max=0, post_max=0` is a valid input, but this is incompatible with point 1 above, since that defines an empty slice.  Similarly for `*_avg`.

Relatedly, I'm not sure `post_*=0` makes sense, since point 1 requires that `x[n]` be a member of the comparison window.  (`pre_*=0` is fine because of the half-open interval semantics of ndarray indexing.)

I think we should either:
  a. require strictly positive `post_*` parameters, or
  b. explicitly add 1 onto the slice and change the guarantees in the docstring.

What do you say?

**Comment by craffel** [comment]
> a. require strictly positive post_* parameters
> 

This seems fine to me.

**Comment by bmcfee** [comment]
Ok, I've made the changes locally.  I'll push when the tests succeed properly.

**Comment by bmcfee** [comment]
@craffel I'm getting failures on peak_pick condition 2; it seems to be picking points that are less than delta above the local mean.  In an extreme case (len(x) == 1, minimal pre/post/wait settings), I get

    x[n] = 1.76950448645
    x[n - pre_avg:n + post_avg] = 1.76950448645 
    delta = 0.5

according to the definition, this signal should have no peaks.  It looks like this is getting thrown by zero-padding x?  I'm thinking edge-padding makes more sense here.  What do you say?

**Comment by bmcfee** [comment]
Switching to `mode=nearest` fixes many of the test failures, but not all of them (including non-trivial inputs).

Upon closer inspection, it looks like this is coming down to the difference between `uniform1d` and taking the mean on slices around the peaks.

**Comment by craffel** [comment]
Hmm.  I think `constant` make sense for `maximum_filter1d` and `mirror` makes sense for `uniform_filter1d`.  Does that cause issues?  What are you comparing against?

**Comment by bmcfee** [comment]
I buy `constant` for max, if `cval = x.min()`.

Mirroring makes me nervous here, since this introduces a dependency forward in time for the average.  This will impact the greedy picking down the line.

My test suite generates random non-negative (`x = np.random.randn(n)**2`) input, and sweeps over all parameters (both valid and invalid) configurations.

I was getting some issues with the "avg_length" and "max_length" parameters; adding 1 is incorrect here.  They should just be the sum of `pre_` and `post_`.  Fixing that cleaned up a lot of the test errors.  This change means that we'll have to redo the parameter sweep for onset detection, sorry!

I'm now back on padding issues, and hitting this case:

```python
>>> x = np.asarray([0.09093222312665819, 0.039408068860174061, 0.35888806955525482, 0.93104182110753808, 0.00094867567386567404])
>>> pre_max, post_max, pre_avg, post_avg, delta, wait = 0, 1, 0, 10, 0.0, 5
>>> librosa.util.peak_pick(x, pre_max, post_max, pre_avg, post_avg, delta, wait)
array([2])
>>> n = peaks[0]
>>> np.mean(x[n-pre_avg:n+post_avg]) + delta
0.4302928554455529
>>> x[n]
0.35888806955525482
>>> x[n] >= np.mean(x[n-pre_avg:n+post_avg]) + delta
False
```

This must be due to the filter hallucinating data, rather than trimming the width of the averaging operator.

**Comment by craffel** [comment]
Err, yeah I think `nearest` for `uniform_filter1d` makes sense.  What edge mode is the example you just posted using?  I think it's fine to specify in the docstring what the edge padding being used is, as a separate condition, and leave it at that.

**Comment by bmcfee** [comment]
That example was using 'nearest', which doesn't actually appear to be clearly documented anywhere?  Edge padding isn't an option in ndimage.

Either way, the answer looks wrong, since it should have picked 3 (0.931) rather than 2 (0.359).

I'm thinking we may be better off padding and trimming manually.

Incidentally, are you sure about the origin logic?  The current code does `int(0.5 * (pre_avg - post_avg))` gives 0 (rounds up), but `(pre_avg - post_avg) // 2` would give -1 (floors).

**Comment by craffel** [comment]
> That example was using 'nearest', which doesn't actually appear to be clearly documented anywhere? 

Yeah, I don't see anywhere that the `mode` parameter values are documented.

> Edge padding isn't an option in ndimage.

Huh?  That's what some of the mode values do, I think.

> I'm thinking we may be better off padding and trimming manually.

You're probably right.

> Incidentally, are you sure about the origin logic?

In my initial commit, I was using an explicit floor.
https://github.com/bmcfee/librosa/commit/f0fa392fb6ebebc0bc1774a6651ec7b9cb3eddef
I'm not sure when it changed.  At any rate, I generally trust you more than myself circa 2013.

**Comment by bmcfee** [comment]
> In my initial commit, I was using an explicit floor.

ok, so the origin should be -1 and not 0?

also, disregard my previous comment about wrong answers: i had a brainfart confusing max with avg.

**Comment by bmcfee** [comment]
scratch that, -1 isn't valid.

**Comment by bmcfee** [comment]
e537ec2abcb51272aeba28ad542c5f1159f01d28 includes a more realistic set of test criteria, and adds some numerical stability to get around comparison tests near 0.

We're still failing a handful of tests though: my latest run had 22 failures out of 12288 tests.  @craffel please poke through this when you get a chance.

**Comment by craffel** [comment]
I'm not totally following what the tests are doing.  In both tests it sets
```Python
            i_max = i + pre_max
            s = i_max - pre_max
            t = i_max + post_max
```
Won't `s` always be `i` then?

**Comment by bmcfee** [comment]
> Won't s always be i then?

Yep.  It's just there to declutter the slice on the padded array.

**Comment by craffel** [comment]
Shouldn't `s` be `i - pre_max` (or 0 when `pre_max > i`?

**Comment by craffel** [comment]
This `peak_pick`, which explicitly computes the sliding mean/max in a for loop, does what I expect it to:
https://gist.github.com/craffel/496329b72e5969417257#file-gistfile1-py
This is the test I'm using, which tests how I interpret the definition of the peak picker:
https://gist.github.com/craffel/496329b72e5969417257#file-gistfile2-py
(note that I'm computing `s` and `t` differently than you)

This "explicit" peak pick indeed does pass this test, and the existing one doesn't.  Running
```Python
%timeit peak_pick(np.random.randn(1000), 10, 10, 10, 10, .5, 10)
```
on this `peak_pick` vs the one currently in `librosa` suggests it's about 100 times slower for these particular parameter settings.  (sidenote - I bet if we used numba it'd be faster than the scipy implementation, but I don't think we need to add a numba depedency...)

I've spent some time seeing if I can get the *filter_1d functions to reproduce this behavior, but I don't thin any of their edge modes will do so.  So, we either need to explicitly define edge behavior in the docstring based on whichever edge mode makes most sense to us, or we need to use the slower version.

Does this all check out with your understanding of the issue?

**Comment by bmcfee** [comment]
> I've spent some time seeing if I can get the *filter_1d functions to reproduce this behavior, but I don't thin any of their edge modes will do so.

Does it fail both max and mean tests?  Or just the mean test?

> So, we either need to explicitly define edge behavior in the docstring based on whichever edge mode makes most sense to us, or we need to use the slower version.

I'd rather not use the slower version.

I think we can significantly accelerate your explicit version by using `util.frame` to make windows in-place and computing max/mean along the vertical axis without looping.  It probably still won't be as fast as the ndimage.filter version, but it might be acceptable.


> (sidenote - I bet if we used numba it'd be faster than the scipy implementation, but I don't think we need to add a numba depedency...)

See #157 , planning for 0.5.

**Comment by craffel** [comment]
Created PR #163 which hopefully resolves this.

**Comment by bmcfee** [comment]
Resolved by #163
